### PR TITLE
feature: Add Cloud Optimized GeoTIFF (COG) sample

### PR DESCRIPTION
**Pull request by ketourneau**
## Description
We need to display Cloud Optimized GeoTIFF ([COG](https://www.cogeo.org/)) with iTowns.
A [pull request](https://github.com/iTowns/itowns/pull/1307) already exist about COG but it seems to be abandoned.
We tried to used PR but we got some trouble with UTIF library to parse COG.
We think [geotiff](https://github.com/geotiffjs/geotiff.js/) is a better library to parse COG.

## Motivation and Context
We need to display heavy georeferenced tif file (geotiff) which come from CAD software.
We think the best solution is to convert geotiff to COG and display it with iTowns.
Maybe it can help [this discussion](https://github.com/iTowns/itowns/discussions/2181) ?

We will change the sample url if [this PR](https://github.com/iTowns/iTowns2-sample-data/pull/17) on itowns data is accepted (or maybe we don't need to change the sample url ?).

### CODE REVIEW

**Comment by Pourfex** [comment]
We only support COG that are generated throught this command :

gdal_translate input.tif output.tif -of COG -co COMPRESS=LZW

This means that we need a geotiff (with CRS) and we only support LZW compression for now. This shouldn't be too hard to support other compression like JPEG (with YUV data) or deflate..

File: examples/js/plugins/COGParser.js [hunk 1]
```diff
@@ -0,0 +1,306 @@
+/* global itowns, THREE */
+
+/**
+ * Select the best overview level (or the final image) to match the
+ * requested extent and pixel width and height.
+ *
+ * @param {Source} source The COGSource
```
Comment (Desplandis) [inline examples/js/plugins/COGParser.js]: Add implicit type parameters for correct typechecking:
```suggestion
 * @param {Object} source The COGSource
 * @param {Extent} source.extent
 * @param {GeoTIFFLevel[]} source.levels
 * @param {THREE.Vector2} source.dimensions
```

You could define `GeoTIFFLevel` as:
```js
/**
 * @typedef  {Object} GeoTIFFLevel
 * @property {GeoTIFFImage} image
 * @property {number} width
 * @property {number} height
 * @property {number[]} resolution
 */
```

File: examples/js/plugins/COGParser.js [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+ * @param {Extent} requestExtent The window extent.
+ * @param {number} requestWidth The pixel width of the window.
+ * @param {number} requestHeight The pixel height of the window.
+ * @returns {Object} The selected zoom level.
```
Comment (Desplandis) [inline examples/js/plugins/COGParser.js]:
```suggestion
 * @returns {GeoTIFFLevel} The selected zoom level.
```

File: examples/js/plugins/COGParser.js [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+        }
+    }
+
+    return level;
+}
+
+/**
+ * Returns a window in the image's coordinates that matches the requested extent.
+ *
+ * @param {Source} source The COGSource
```
Comment (Desplandis) [inline examples/js/plugins/COGParser.js]: Add implicit type parameters for correct typechecking:
```suggestion
 * @param {Object} source The COGSource
 * @param {number[]} source.origin Root image origin as an XYZ-vector
```

File: examples/js/plugins/COGParser.js [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+    }
+
+    return [xMin, yMin, xMax, yMax];
+}
+
+/**
+ * Creates a texture from the pixel buffer(s).
+ *
+ * @param {Source} source The COGSource
+ * @param {Object} buffers The buffers (one buffer per band)
```
Comment (Desplandis) [inline examples/js/plugins/COGParser.js]: Add implicit type parameters for correct typechecking:
```suggestion
 * @param {THREE.TypedArray | THREE.TypedArray[]} buffers The buffers (one buffer per band)
 * @param {number} buffers.width
 * @param {number} buffers.height
```

File: examples/js/plugins/COGParser.js [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+    return texture;
+}
+
+
+// Important note : a lot of code is duplicated to avoid putting
+// conditional branches inside loops, as this can severely reduce performance.
+// Note: we don't use Number.isNan(x) in the loops as it slows down the loop due to function
+// invocation. Instead, we use x !== x, as a NaN is never equal to itself.
+/* eslint no-self-compare: 0 */
+function fillBuffer(source, buffers, options, opaqueValue, pixelData) {
```
Comment (Desplandis) [inline examples/js/plugins/COGParser.js]: I think we could reduce the number of parameters:
- the parameters with different parameters between calls (i.e. `pixelData` and `buffers`) could be passed as first parameters
- the properties of `source` used here (i.e. `defaultNoData` and `defaultTransparent`) could be passed as a unique `options` object.

```suggestion
function fillBuffer(pixelData, buffers, options) {
```

File: examples/js/plugins/COGParser.js [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+                value = source.defaultNoData;
+                a = source.defaultTransparent;
+            } else {
+                value = raw;
+                a = opaqueValue;
+            }
+            buffers[idx + 0] = value;
+            buffers[idx + 1] = a;
+        }
+    }
```
Comment (Desplandis) [inline examples/js/plugins/COGParser.js]: What is the intended behavior of this code? Given the above comment, the grey channel will be interpreted as the red color and the default transparency as the green color.

Dataset with a unique channel (from [OpenAerialMap](https://map.openaerialmap.org/)):
https://oin-hotosm.s3.amazonaws.com/60fbca155a90f10006fd2fc3/0/60fbca155a90f10006fd2fc4.tif

File: examples/js/plugins/COGParser.js [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+            const raw = v[i];
+            if (raw !== raw || raw === options.noData) {
+                value = source.defaultNoData;
+            } else {
+                value = raw;
+            }
+            buffers[idx + 0] = value;
+            buffers[idx + 1] = a[i];
+        }
+    }
```
Comment (Desplandis) [inline examples/js/plugins/COGParser.js]: Same comment and question as the previous comment.

File: examples/js/plugins/COGParser.js [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+                b = getValue(b);
+                a = opaqueValue;
+            }
+
+            buffers[idx + 0] = r;
+            buffers[idx + 1] = g;
+            buffers[idx + 2] = b;
+            buffers[idx + 3] = a;
+        }
+    }
```
Comment (Desplandis) [inline examples/js/plugins/COGParser.js]: There seems to be an issue with an orthos from IGN's [PCRS](https://geoservices.ign.fr/pcrs).
You can download it from `ftp://JeuTest_Ouroux:Eigoo9ni8keefi8t@ftp3.ign.fr/`.

File: examples/js/plugins/COGParser.js [hunk 9]
```diff
@@ -0,0 +1,306 @@
...
+        THREE = itowns.THREE;
+    }
+
+    return {
+        /**
+         * Parse a COG file and return a `THREE.DataTexture`.
+         *
+         * @param {Object} data Data passed with the Tile extent
+         * @param {Object} options Options (contains source)
+         * @return {Promise} A promise resolving with a `THREE.DataTexture`.
```
Comment (Desplandis) [inline examples/js/plugins/COGParser.js]: Missing types of parameters:

```suggestion
         * @param {Object} data Data passed with the Tile extent
         * @param {Extent} data.extent
         * @param {Object} options Options (contains source)
         * @param {Object} options.in
         * @param {COGSource} options.in.source
         * @param {number} options.in.tileWidth
         * @param {number} options.in.tileHeight
         * @return {Promise<THREE.DataTexture>} A promise resolving with a `THREE.DataTexture`.
```

File: examples/js/plugins/COGParser.js [hunk 10]
```diff
@@ -0,0 +1,306 @@
...
+         *
+         * @memberof module:COGParser
+         */
+        parse: async function _(data, options) {
+            const source = options.in;
+            const nodeExtent = data.extent.as(source.crs);
+            const level = selectLevel(source, nodeExtent, source.tileWidth, source.tileHeight);
+            const viewport = makeWindowFromExtent(source, nodeExtent, level.resolution);
+
+            const buffers = await level.image.readRasters({
```
Comment (Desplandis) [inline examples/js/plugins/COGParser.js]: There seems to be a way to directly load a RGB from certain types of `geotiff` which returns 3 channels as `TypedArray[]` (see `GeotiffImage#readRGB`). Maybe this could simplify this PR and solves the issues above?

File: examples/js/plugins/COGSource.js [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+
+        this.whenReady = GeoTIFF.fromUrl(this.url)
+            .then(async (geotiff) => {
+                this.geotiff = geotiff;
+                this.firstImage = await geotiff.getImage();
+                this.origin = this.firstImage.getOrigin();
+                this.noData = (source.noData !== undefined) ? source.noData : this.firstImage.getGDALNoData();
+                this.dataType = this.selectDataType(this.firstImage.getSampleFormat(), this.firstImage.getBitsPerSample());
+
+                this.channels = [];
```
Comment (Desplandis) [inline examples/js/plugins/COGSource.js]: This could be the object of another PR but we could parameter which channels to show and their order.

File: examples/js/plugins/COGParser.js [hunk 12]
```diff
@@ -0,0 +1,306 @@
...
+/* eslint no-self-compare: 0 */
+function fillBuffer(source, buffers, options, opaqueValue, pixelData) {
+    let getValue;
+
+    if (options.scaling !== undefined) {
+        const { min, max } = options.scaling;
+        getValue = x => Math.floor(THREE.MathUtils.mapLinear(x, min, max, 0, 255));
+    } else {
+        getValue = x => x;
+    }
```
Comment (Desplandis) [inline examples/js/plugins/COGParser.js]: This `scaling` option is never passed to `fillBuffer`. You should either remove this code or pass the option to `fillBuffer` calls.

File: examples/js/plugins/COGSource.js [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+ *
+ * // Add the layer
+ * view.addLayer(colorLayer);
+ */
+class COGSource extends itowns.Source {
+    /**
+     * @param {Object} source - An object that can contain all properties of a
+     * COGSource and {@link Source}. Only `url` is mandatory.
+     *
+     * @constructor
```
Comment (Desplandis) [inline examples/js/plugins/COGSource.js:37]: Missing JSDOC for `source.zoom`, `source.url`, `source.pool` (of type `GeoTIFF.Pool`), `source.defaultNoData`, `source.defaultTransparent`, `source.opaqueByte` and `source.opaqueFloat`.

File: examples/js/plugins/COGSource.js [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+        this.url = source.url;
+        this.pool = source.pool || new GeoTIFF.Pool();
+        // We don't use fetcher, we let geotiff.js manage it
+        this.fetcher = () => Promise.resolve({});
+        this.parser = COGParser.parse;
+
+        this.defaultNoData = source.defaultNoData || 255;
+        this.defaultTransparent = source.defaultTransparent || 0;
+        this.opaqueByte = source.opaqueByte || 255;
+        this.opaqueFloat = source.opaqueFloat || 1;
```
Comment (Desplandis) [inline examples/js/plugins/COGSource.js]: Could we not merge those two options ? `opaqueFloat` seems to be `opaqueByte / 255`.

File: examples/js/plugins/COGSource.js [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+
+        this.whenReady = GeoTIFF.fromUrl(this.url)
+            .then(async (geotiff) => {
+                this.geotiff = geotiff;
+                this.firstImage = await geotiff.getImage();
+                this.origin = this.firstImage.getOrigin();
+                this.noData = (source.noData !== undefined) ? source.noData : this.firstImage.getGDALNoData();
```
Comment (Desplandis) [inline examples/js/plugins/COGSource.js]:
```suggestion
                this.noData = source.noData ?? this.firstImage.getGDALNoData();
```

File: examples/js/plugins/COGSource.js [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+                this.dataType = this.selectDataType(this.firstImage.getSampleFormat(), this.firstImage.getBitsPerSample());
+
+                this.channels = [];
+                for (let i = 0; i < this.firstImage.getSamplesPerPixel(); i += 1) {
+                    this.channels.push(i);
```
Comment (Desplandis) [inline examples/js/plugins/COGSource.js]: This could lead to some issue with `geotiff` with `mask` channels. We could maybe ignore them for now?
See https://www.awaresystems.be/imaging/tiff/tifftags/newsubfiletype.html

File: examples/js/plugins/COGSource.js [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+                this.dimensions = this.extent.planarDimensions();
+
+                this.levels = [];
+                this.levels.push(this.makeLevel(this.firstImage, this.firstImage.getResolution()));
+
+                // Number of images (original + overviews)
+                const imageCount = await this.geotiff.getImageCount();
+
+                // We want to preserve the order of the overviews so we await them inside
+                // the loop not to have the smallest overviews coming before the biggest
```
Comment (Desplandis) [inline examples/js/plugins/COGSource.js]: This comment seems to not be accurate, [`Promise.all`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all) preservers the order of the array of promises passed (see [`Promise.all#return_value`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all#return_value)). The following code could then be rewritten without `await` in loop.

File: examples/js/plugins/COGSource.js [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+                for (let index = 1; index < imageCount; index++) {
+                    // eslint-disable-next-line no-await-in-loop
+                    const image = await this.geotiff.getImage(index);
+                    this.levels.push(this.makeLevel(image, image.getResolution(this.firstImage)));
+                }
+            });
+    }
+
+    selectDataType(format, bitsPerSample) {
```
Comment (Desplandis) [inline examples/js/plugins/COGSource.js:94]:
```suggestion
    /**
     * @param {number} format - Format to interpret each data sample in a pixel
     * https://www.awaresystems.be/imaging/tiff/tifftags/sampleformat.html
     * @param {number} bitsPerSample - Number of bits per component.
     * https://www.awaresystems.be/imaging/tiff/tifftags/bitspersample.html
     * @return {THREE.AttributeGPUType}
     */
    selectDataType(format, bitsPerSample) {
```

File: examples/source_file_cog.html [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+<script src="../dist/itowns.js"></script>
+<script src="../dist/debug.js"></script>
+<script src="js/GUI/LoadingScreen.js"></script>
+<script src="https://cdn.jsdelivr.net/npm/geotiff"></script>
+<script src="js/plugins/COGParser.js"></script>
+<script src="js/plugins/COGSource.js"></script>
+<script type="text/javascript">
+    itowns.proj4.defs('EPSG:3947', '+proj=lcc +lat_0=47 +lon_0=3 +lat_1=46.25 +lat_2=47.75 +x_0=1700000 +y_0=6200000 +ellps=GRS80 +towgs84=0,0,0,0,0,0,0 +units=m +no_defs +type=crs');
+    var viewerDiv = document.getElementById('viewerDiv');
+
```
Comment (Desplandis) [inline examples/source_file_cog.html]: Could you please add an url parameter (maybe `geotiff`?)? This could ease the request of other files. =)
We could default to your dataset if no such parameter is provided!

File: examples/js/plugins/COGSource.js [hunk 20]
```diff
@@ -0,0 +1,123 @@
...
+        this.opaqueByte = source.opaqueByte || 255;
+        this.opaqueFloat = source.opaqueFloat || 1;
+
+        this.whenReady = GeoTIFF.fromUrl(this.url)
+            .then(async (geotiff) => {
+                this.geotiff = geotiff;
+                this.firstImage = await geotiff.getImage();
+                this.origin = this.firstImage.getOrigin();
+                this.noData = (source.noData !== undefined) ? source.noData : this.firstImage.getGDALNoData();
+                this.dataType = this.selectDataType(this.firstImage.getSampleFormat(), this.firstImage.getBitsPerSample());
```
Comment (Desplandis) [inline examples/js/plugins/COGSource.js:61]: Just leaving a comment but I don't know if each band of a geotiff is of the same datatype.

File: examples/js/plugins/COGParser.js [hunk 21]
```diff
@@ -0,0 +1,306 @@
...
+    const { width, height } = buffers;
+    const pixelCount = width * height;
+    const targetDataType = source.dataType;
+
+    let format;
+    let channelCount;
+    switch (buffers.length) {
+        case 1:
+        case 2:
+            format = THREE.RGFormat;
```
Comment (Desplandis) [inline examples/js/plugins/COGParser.js]: There is no dedicated code in our fragment to interpret `red` and `green` pairs as for example grey and transparency pairs. This could lead to "weird" colors when viewing COG with single or dual band.

**Review by Desplandis — CHANGES_REQUESTED** [review]
Hi and thanks for this PR! =)
In my opinion, this is a must-have feature so it is pretty nice to have started an implementation! I think we will let it as plug-in for now and integrate it in core itowns when the implementation will be mature enough.

I have a number of comments on this PR and found a number of datasets on [OpenArialMaps](https://map.openaerialmap.org/). A lot of them failed to load since they have mask channels however!

File: examples/js/plugins/COGParser.js (hunk 1 above)
Comment (ketourneau) [reply examples/js/plugins/COGParser.js]: Ok, done !

File: examples/js/plugins/COGParser.js (hunk 2 above)
Comment (ketourneau) [reply examples/js/plugins/COGParser.js]: Ok, done !

File: examples/js/plugins/COGParser.js (hunk 3 above)
Comment (ketourneau) [reply examples/js/plugins/COGParser.js]: Ok, done !

File: examples/js/plugins/COGParser.js (hunk 21 above)
Comment (ketourneau) [reply examples/js/plugins/COGParser.js]: Ok we alaways consider there are 4 channel

File: examples/js/plugins/COGParser.js (hunk 9 above)
Comment (ketourneau) [reply examples/js/plugins/COGParser.js]: Ok, done

File: examples/js/plugins/COGSource.js (hunk 14 above)
Comment (ketourneau) [reply examples/js/plugins/COGSource.js]: Yes indeed, I made the change.

File: examples/js/plugins/COGSource.js (hunk 15 above)
Comment (ketourneau) [reply examples/js/plugins/COGSource.js]: Ok, done !

File: examples/js/plugins/COGSource.js (hunk 18 above)
Comment (ketourneau) [reply examples/js/plugins/COGSource.js:94]: Ok done

File: examples/js/plugins/COGParser.js (hunk 12 above)
Comment (ketourneau) [reply examples/js/plugins/COGParser.js]: Yes indeed I remove the scaling option

File: examples/js/plugins/COGParser.js (hunk 5 above)
Comment (ketourneau) [reply examples/js/plugins/COGParser.js]: Ok but we need to let opaqueValue as parameter (there are 2 possible values), it's ok for you ?

File: examples/js/plugins/COGParser.js (hunk 6 above)
Comment (ketourneau) [reply examples/js/plugins/COGParser.js]: We've modified it to fill a RGBA color

File: examples/js/plugins/COGParser.js (hunk 8 above)
Comment (ketourneau) [reply examples/js/plugins/COGParser.js]: By default GDAL generate YUV COG, actually we only support LZW

File: examples/js/plugins/COGSource.js (hunk 13 above)
Comment (ketourneau) [reply examples/js/plugins/COGSource.js:37]: Ok done

File: examples/js/plugins/COGSource.js (hunk 20 above)
Comment (ketourneau) [reply examples/js/plugins/COGSource.js:61]: That's a good question, Do you have an example of data to test ?

File: examples/source_file_cog.html (hunk 19 above)
Comment (ketourneau) [reply examples/source_file_cog.html]: Ok done

File: examples/js/plugins/COGSource.js (hunk 16 above)
Comment (ketourneau) [reply examples/js/plugins/COGSource.js]: Ok done

File: examples/js/plugins/COGSource.js (hunk 11 above)
Comment (ketourneau) [reply examples/js/plugins/COGSource.js]: Yes, indeed. Actually we ignore channels to prevent some bugs.

File: examples/js/plugins/COGParser.js (hunk 10 above)
Comment (ketourneau) [reply examples/js/plugins/COGParser.js]: Ok we try to use Geotiff readRGB with alpha support and it's work in our case !

**Review by ketourneau — COMMENTED** [review]
@Desplandis I apply your code review, can you check if everything is good for you ?

File: examples/js/plugins/COGSource.js (hunk 20 above)
Comment (Desplandis) [reply examples/js/plugins/COGSource.js:61]: No example for now!

**Review by Desplandis — APPROVED** [review]
Ok to me!
